File: MapsSDKExamplesObjc/Routing/RoutingBatchRouteViewController.m
/**
 * Copyright (c) 2019 TomTom N.V. All rights reserved.
 *
 * This software is the proprietary copyright of TomTom N.V. and its
 * subsidiaries and may be used for internal evaluation purposes or commercial
 * use strictly subject to separate licensee agreement between you and TomTom.
 * If you are the licensee, you are only permitted to use this Software in
 * accordance with the terms of your license agreement. If you are not the
 * licensee then you are not authorised to use this software in any manner and
 * should immediately return it to TomTom N.V.
 */

#import "RoutingBatchRouteViewController.h"

typedef NS_ENUM(NSInteger, BatchRouteType) {
  BatchRouteTypeTravelMode,
  BatchRouteTypeRoute,
  BatchRouteTypeAvoids
} BatchRouteSectionType;

@interface RoutingBatchRouteViewController () <
    TTBatchRouteResponseDelegate, TTBatchRouteVisistor, TTRouteDelegate>

@property(nonatomic, strong) TTBatchRoute *batchRoute;
@property NSDictionary *routeDesc;
@property BatchRouteType type;

@end

@implementation RoutingBatchRouteViewController

- (void)setupDict {
  self.routeDesc = @{
    @(BatchRouteTypeTravelMode) :
        @[ @"Travel by Car", @"Travel by Truck", @"Travel by Pedestrian" ],
    @(BatchRouteTypeRoute) :
        @[ @"Fastest route", @"Shortest route", @"Eco route" ],
    @(BatchRouteTypeAvoids) :
        @[ @"Avoid motorways", @"Avoid ferries", @"Avoid toll roads" ]
  };
}

- (OptionsView *)getOptionsView {
  return [[OptionsViewSingleSelect alloc]
      initWithLabels:@[ @"Travel mode", @"Route type", @"Avoids" ]
          selectedID:-1];
}

- (void)setupCenterOnWillHappen {
  [self.mapView centerOnCoordinate:[TTCoordinate AMSTERDAM] withZoom:10];
}

- (void)viewDidLoad {
  [super viewDidLoad];
  [self setupDict];
  self.mapView.routeManager.delegate = self;
  self.batchRoute = [TTBatchRoute new];
  self.batchRoute.delegate = self;
}

#pragma mark OptionsViewDelegate

- (void)displayExampleWithID:(NSInteger)ID on:(BOOL)on {
  [super displayExampleWithID:ID on:on];
  [self.mapView.routeManager removeAllRoutes];
  [self.progress show];
  [self.etaView hide];
  switch (ID) {
  case 2:
    [self performAvoids];
    break;
  case 1:
    [self performRouteType];
    break;
  default:
    [self performTravelMode];
    break;
  }
}

#pragma mark Examples

- (void)performTravelMode {
  self.type = BatchRouteTypeTravelMode;
  TTRouteQuery *queryCar = [[[[[TTRouteQueryBuilder
      createWithDest:[TTCoordinate ROTTERDAM]
             andOrig:[TTCoordinate AMSTERDAM]] withComputeBestOrder:YES]
      withTraffic:YES] withTravelMode:TTOptionTravelModeCar] build];
  TTRouteQuery *queryTruck = [[[[[TTRouteQueryBuilder
      createWithDest:[TTCoordinate ROTTERDAM]
             andOrig:[TTCoordinate AMSTERDAM]] withComputeBestOrder:YES]
      withTraffic:YES] withTravelMode:TTOptionTravelModeTruck] build];
  TTRouteQuery *queryPedestrain = [[[[[TTRouteQueryBuilder
      createWithDest:[TTCoordinate ROTTERDAM]
             andOrig:[TTCoordinate AMSTERDAM]] withComputeBestOrder:YES]
      withTraffic:YES] withTravelMode:TTOptionTravelModePedestrian] build];
  TTBatchRouteQuery *batchQuery =
      [[[[TTBatchRouteQueryBuilder createRouteQuery:queryCar]
          addRouteQuery:queryTruck] addRouteQuery:queryPedestrain] build];

  self.batchRoute = [[TTBatchRoute alloc] init];
  self.batchRoute.delegate = self;
  [self.batchRoute batchRouteWithQuery:batchQuery];
}

- (void)performRouteType {
  self.type = BatchRouteTypeRoute;
  TTRouteQuery *queryFastest = [[[[[TTRouteQueryBuilder
      createWithDest:[TTCoordinate ROTTERDAM]
             andOrig:[TTCoordinate AMSTERDAM]] withComputeBestOrder:YES]
      withTraffic:YES] withRouteType:TTOptionTypeRouteFastest] build];
  TTRouteQuery *queryShortest = [[[[[TTRouteQueryBuilder
      createWithDest:[TTCoordinate ROTTERDAM]
             andOrig:[TTCoordinate AMSTERDAM]] withComputeBestOrder:YES]
      withTraffic:YES] withRouteType:TTOptionTypeRouteShortest] build];
  TTRouteQuery *queryEco = [[[[[TTRouteQueryBuilder
      createWithDest:[TTCoordinate ROTTERDAM]
             andOrig:[TTCoordinate AMSTERDAM]] withComputeBestOrder:YES]
      withTraffic:YES] withRouteType:TTOptionTypeRouteEco] build];

  TTBatchRouteQuery *batchQuery =
      [[[[TTBatchRouteQueryBuilder createRouteQuery:queryFastest]
          addRouteQuery:queryShortest] addRouteQuery:queryEco] build];
  self.batchRoute = [[TTBatchRoute alloc] init];
  self.batchRoute.delegate = self;
  [self.batchRoute batchRouteWithQuery:batchQuery];
}

- (void)performAvoids {
  self.type = BatchRouteTypeAvoids;
  TTRouteQuery *queryMotorways = [[[[[TTRouteQueryBuilder
      createWithDest:[TTCoordinate OSLO]
             andOrig:[TTCoordinate AMSTERDAM]] withComputeBestOrder:YES]
      withTraffic:YES] withAvoidType:TTOptionTypeAvoidMotorways] build];
  TTRouteQuery *queryFerries = [[[[[TTRouteQueryBuilder
      createWithDest:[TTCoordinate OSLO]
             andOrig:[TTCoordinate AMSTERDAM]] withComputeBestOrder:YES]
      withTraffic:YES] withAvoidType:TTOptionTypeAvoidFerries] build];
  TTRouteQuery *queryTollRoads = [[[[[TTRouteQueryBuilder
      createWithDest:[TTCoordinate OSLO]
             andOrig:[TTCoordinate AMSTERDAM]] withComputeBestOrder:YES]
      withTraffic:YES] withAvoidType:TTOptionTypeAvoidTollRoads] build];

  TTBatchRouteQuery *batchQuery =
      [[[[TTBatchRouteQueryBuilder createRouteQuery:queryMotorways]
          addRouteQuery:queryFerries] addRouteQuery:queryTollRoads] build];
  self.batchRoute = [[TTBatchRoute alloc] init];
  self.batchRoute.delegate = self;
  [self.batchRoute batchRouteWithQuery:batchQuery];
}

#pragma mark TTBatchRouteResponseDelegate

- (void)batch:(TTBatchRoute *_Nonnull)route
    completedWithResponse:(TTBatchRouteResponse *_Nonnull)response {
  [self.progress hide];
  [response visit:self];
}

- (void)batch:(TTBatchRoute *_Nonnull)route
    failedWithError:(TTResponseError *_Nonnull)responseError {
  [self handleError:responseError];
}

#pragma mark TTBatchRouteVisistor

- (void)visitRoute:(TTRouteResult *)response {
  TTMapRoute *mapRoute =
      [TTMapRoute routeWithCoordinatesData:response.routes.firstObject
                            withRouteStyle:TTMapRouteStyle.defaultInactiveStyle
                                imageStart:TTMapRoute.defaultImageDeparture
                                  imageEnd:TTMapRoute.defaultImageDestination];
  mapRoute.extraData = response.routes.firstObject.summary;
  [self.mapView.routeManager addRoute:mapRoute];
  [self.progress hide];
  [self.mapView.routeManager showAllRoutesOverview];
}

#pragma mark TTRouteDelegate

- (void)routeClicked:(TTMapRoute *_Nonnull)route {
  NSString *desc;
  for (TTMapRoute *mapRoute in self.mapView.routeManager.routes) {
    [self.mapView.routeManager
        updateRoute:mapRoute
              style:TTMapRouteStyle.defaultInactiveStyle];
  }
  [self.mapView.routeManager updateRoute:route
                                   style:TTMapRouteStyle.defaultActiveStyle];
  [self.mapView.routeManager bringToFrontRoute:route];
  desc = [(NSArray *)[self.routeDesc objectForKey:@(self.type)]
      objectAtIndex:[self.mapView.routeManager.routes indexOfObject:route]];
  [self updateETA:route desc:desc];
}

#pragma Display ETA
- (void)updateETA:(TTMapRoute *)mapRoute desc:(NSString *)desc {
  TTSummary *summary = (TTSummary *)mapRoute.extraData;

  NSDate *eta = summary.arrivalTime;
  NSCalendar *calendar = [NSCalendar currentCalendar];
  [calendar setTimeZone:[NSTimeZone localTimeZone]];
  NSDateComponents *components =
      [calendar components:(NSCalendarUnitHour | NSCalendarUnitMinute)
                  fromDate:eta];
  NSString *dateInString =
      [NSString stringWithFormat:@"%02ld:%02ld %@", components.hour,
                                 components.minute, desc];
  [self.etaView updateWithEta:dateInString
               metersDistance:summary.lengthInMetersValue];
}

@end

File: MapsSDKExamplesSwift/Routing/RoutingBatchRouteViewController.swift
/**
 * Copyright (c) 2019 TomTom N.V. All rights reserved.
 *
 * This software is the proprietary copyright of TomTom N.V. and its subsidiaries and may be used
 * for internal evaluation purposes or commercial use strictly subject to separate licensee
 * agreement between you and TomTom. If you are the licensee, you are only permitted to use
 * this Software in accordance with the terms of your license agreement. If you are not the
 * licensee then you are not authorised to use this software in any manner and should
 * immediately return it to TomTom N.V.
 */

import MapsSDKExamplesCommon
import MapsSDKExamplesVC
import TomTomOnlineSDKMaps
import TomTomOnlineSDKRouting
import UIKit

enum BatchRouteType: Int {
    case BatchRouteTypeTravelMode
    case BatchRouteTypeRoute
    case BatchRouteTypeAvoids
}

class RoutingBatchRouteViewController: RoutingBaseViewController, TTBatchRouteVisistor, TTBatchRouteResponseDelegate, TTRouteDelegate {
    var batchRoute = TTBatchRoute()
    var type = BatchRouteType.BatchRouteTypeTravelMode
    var routeDesc = [BatchRouteType.BatchRouteTypeTravelMode: ["Travel by Car", "Travel by Truck", "Travel by Pedestrian"],
                     BatchRouteType.BatchRouteTypeRoute: ["Fastest route", "Shortest route", "Eco route"],
                     BatchRouteType.BatchRouteTypeAvoids: ["Avoid motorways", "Avoid ferries", "Avoid toll roads"]]

    override func setupInitialCameraPosition() {
        mapView.center(on: TTCoordinate.AMSTERDAM(), withZoom: 10)
    }

    override func getOptionsView() -> OptionsView {
        return OptionsViewSingleSelect(labels: ["Travel mode", "Route type", "Avoids"], selectedID: -1)
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        batchRoute = TTBatchRoute()
        batchRoute.delegate = self
        mapView.routeManager.delegate = self
    }

    // MARK: OptionsViewDelegate

    override func displayExample(withID ID: Int, on: Bool) {
        super.displayExample(withID: ID, on: on)
        mapView.routeManager.removeAllRoutes()
        progress.show()
        etaView.hide()

        switch ID {
        case 2:
            performAvoids()
            type = BatchRouteType.BatchRouteTypeAvoids
        case 1:
            performRouteType()
            type = BatchRouteType.BatchRouteTypeRoute
        default:
            performTravelMode()
            type = BatchRouteType.BatchRouteTypeTravelMode
        }
    }

    // MARK: Examples

    func performTravelMode() {
        type = .BatchRouteTypeTravelMode
        let queryCar = TTRouteQueryBuilder.create(withDest: TTCoordinate.ROTTERDAM(), andOrig: TTCoordinate.AMSTERDAM())
            .withComputeBestOrder(true)
            .withTraffic(true)
            .withTravelMode(TTOptionTravelMode.car)
            .build()

        let queryTruck = TTRouteQueryBuilder.create(withDest: TTCoordinate.ROTTERDAM(), andOrig: TTCoordinate.AMSTERDAM())
            .withComputeBestOrder(true)
            .withTraffic(true)
            .withTravelMode(TTOptionTravelMode.truck)
            .build()

        let queryPedestrain = TTRouteQueryBuilder.create(withDest: TTCoordinate.ROTTERDAM(), andOrig: TTCoordinate.AMSTERDAM())
            .withComputeBestOrder(true)
            .withTraffic(true)
            .withTravelMode(TTOptionTravelMode.pedestrian)
            .build()
        let batchQuery = TTBatchRouteQueryBuilder.createRouteQuery(queryCar)
            .add(queryTruck)
            .add(queryPedestrain)
            .build()

        batchRoute = TTBatchRoute()
        batchRoute.delegate = self
        batchRoute.batchRoute(with: batchQuery)
    }

    func performRouteType() {
        type = .BatchRouteTypeRoute
        let queryFastest = TTRouteQueryBuilder.create(withDest: TTCoordinate.ROTTERDAM(), andOrig: TTCoordinate.AMSTERDAM())
            .withComputeBestOrder(true)
            .withTraffic(true)
            .withRouteType(TTOptionTypeRoute.fastest)
            .build()

        let queryShortest = TTRouteQueryBuilder.create(withDest: TTCoordinate.ROTTERDAM(), andOrig: TTCoordinate.AMSTERDAM())
            .withComputeBestOrder(true)
            .withTraffic(true)
            .withRouteType(TTOptionTypeRoute.shortest)
            .build()

        let queryEco = TTRouteQueryBuilder.create(withDest: TTCoordinate.ROTTERDAM(), andOrig: TTCoordinate.AMSTERDAM())
            .withComputeBestOrder(true)
            .withTraffic(true)
            .withRouteType(TTOptionTypeRoute.eco)
            .build()

        let batchQuery = TTBatchRouteQueryBuilder.createRouteQuery(queryFastest)
            .add(queryShortest)
            .add(queryEco)
            .build()

        batchRoute = TTBatchRoute()
        batchRoute.delegate = self
        batchRoute.batchRoute(with: batchQuery)
    }

    func performAvoids() {
        type = .BatchRouteTypeRoute
        let queryMotorways = TTRouteQueryBuilder.create(withDest: TTCoordinate.OSLO(), andOrig: TTCoordinate.AMSTERDAM())
            .withComputeBestOrder(true)
            .withTraffic(true)
            .withAvoidType(TTOptionTypeAvoid.motorways)
            .build()

        let queryFerries = TTRouteQueryBuilder.create(withDest: TTCoordinate.OSLO(), andOrig: TTCoordinate.AMSTERDAM())
            .withComputeBestOrder(true)
            .withTraffic(true)
            .withAvoidType(TTOptionTypeAvoid.ferries)
            .build()

        let queryTollRoads = TTRouteQueryBuilder.create(withDest: TTCoordinate.OSLO(), andOrig: TTCoordinate.AMSTERDAM())
            .withComputeBestOrder(true)
            .withTraffic(true)
            .withAvoidType(TTOptionTypeAvoid.tollRoads)
            .build()

        let batchQuery = TTBatchRouteQueryBuilder.createRouteQuery(queryMotorways)
            .add(queryFerries)
            .add(queryTollRoads)
            .build()

        batchRoute = TTBatchRoute()
        batchRoute.delegate = self
        batchRoute.batchRoute(with: batchQuery)
    }

    // MARK: TTBatchRouteResponseDelegate

    func batch(_: TTBatchRoute, completedWith response: TTBatchRouteResponse) {
        progress.hide()
        response.visit(self)
    }


    func batch(_: TTBatchRoute, failedWithError responseError: TTResponseError) {
        handleError(responseError)
    }


    // MARK: TTBatchRouteVisistor

    func visitRoute(_ response: TTRouteResult) {
        let mapRoute = TTMapRoute(coordinatesData: response.routes.first!,
                                  with: TTMapRouteStyle.defaultInactive(),
                                  imageStart: TTMapRoute.defaultImageDeparture(),
                                  imageEnd: TTMapRoute.defaultImageDestination())
        mapRoute.extraData = response.routes.first?.summary
        mapView.routeManager.add(mapRoute)
        progress.hide()
        displayRouteOverview()
    }


    // MARK: TTRouteDelegate

    func routeClicked(_ route: TTMapRoute) {
        for mapRoute in self.mapView.routeManager.routes {
            mapView.routeManager.update(mapRoute, style: TTMapRouteStyle.defaultInactive())
        }
        mapView.routeManager.update(route, style: TTMapRouteStyle.defaultActive())
        mapView.routeManager.bring(toFrontRoute: route)
        let desc = routeDesc[type]![mapView.routeManager.routes.index(of: route)!]
        updateEta(mapRoute: route, desc: desc)
    }

    // MARK: display ETA

    func updateEta(mapRoute: TTMapRoute, desc: String) {
        let summary = mapRoute.extraData as! TTSummary
        let eta = summary.arrivalTime

        let calendar = Calendar.current
        let unitFlags = Set<Calendar.Component>([.hour, .minute])
        let components: NSDateComponents = calendar.dateComponents(unitFlags, from: eta) as NSDateComponents
        let dateInString = "\(components.hour):\(components.minute) \(desc)"
        etaView.update(eta: dateInString, metersDistance: UInt(summary.lengthInMetersValue))
    }
}
